### DISCUSSION TITLE
upstreams cache invalidate_local

### DISCUSSION THREAD

**Discussion by liukaiyuliukaiyu** [Kong/kong#11563] [Help]
Hi,if one target of a service is updated, or post a new target, or delete a old target,the whole upstreams cache will be invalidated, Why invalidate the entire upstreams, rather than just invalidate the upstreams for each service you operate on？
Can it be optimized?
Extract the value associated with the cache key"upstreams", clear the upstream_name and upstream_id associated with the value, and set it back to shared memory. 
Extract the value of the cache key"upstreams", add the corresponding upstream_name and upstream_id pairs to the value, and set it back to shared memory.
I would appreciate it if you could help me look at this problem@bungle  @kikito @hbagdi 
The following code snippet：

 if operation == "create" then

    singletons.cache:invalidate_local("balancer:upstreams")

    local _, err = create_balancer(upstream)
    if err then
      log(ERR, "failed creating balancer for ", upstream.name, ": ", err)
    end

  elseif operation == "delete" or operation == "update" then

    singletons.cache:invalidate_local("balancer:upstreams")
    singletons.cache:invalidate_local("balancer:upstreams:" .. upstream.id)
    singletons.cache:invalidate_local("balancer:targets:"   .. upstream.id)

**Comment by liukaiyuliukaiyu** [comment]
"balancer:upstreams"` - a list of upstreams
--to be invalidated on any upstream change
Why update an upstream and invalidate all upstreams?

**Comment by liukaiyuliukaiyu** [comment]
Why not update like this?
I would appreciate it if you could help me look at this problem@bungle  @kikito  @hbagdi 
```
function _M.invalidate_individual_upstream_from_upstreams_dict(upstream_name, upstream_id)
    local upstreams_dict = _M.get(CACHE_KEYS.UPSTREAMS)
    if not upstreams_dict then
        return
    end

    for _, up in ipairs(upstreams_dict) do
        if up.name == upstream_name and upstream_id == up.id then
            upstreams_dict[upstream_name] = nil
            break
        end
    end

    local ok, err = _M.set(CACHE_KEYS.UPSTREAMS, upstreams_dict, nil)
end

function _M.add_individual_upstream_to_upstreams_dict(upstream_name, upstream_id)
    local upstreams_dict = _M.get(CACHE_KEYS.UPSTREAMS)
    if not upstreams_dict then
        return
    end

    upstreams_dict[upstream_name] = upstream_id

    local ok, err = _M.set(CACHE_KEYS.UPSTREAMS, upstreams_dict, nil)
end
```

**Comment by sabertobihwy** [comment]
> Why invalidate the entire upstreams, rather than just invalidate the upstreams for each service you operate on？

I think you misunderstand the relationship between service and upstream here. If you create an upstream and associate some services with it, the services do not have their own copies of this upstream, but instead, they share the configuration of this upstream.

**Reply by liukaiyuliukaiyu** [reply]
I feel like we're talking about two different issues, Maybe the version 0.10.3 we're using is older,However, according to the database relationship, one api will only be associated with one upstream, and one upstream will have multiple targets. The nginx cache structure is as follows:
Let's say we have three apis：api1, api2, api3，
and the corresponding upstream_id is upstream_id1, upstream_id2, upstream_id3，
the lua shared dict cache structure of upstreams is: upstreams={api1=upstream_id1, api2=upstream_id2, api3=upstream_id3}, When delete api1, kong just clear the entire upstreams,in other words, upstreams=nil，But we're not operating the api2 and api3，Whether we can do this，upstreams={api2=upstream_id2, api3=upstream_id3}, rewrite the shared memory, reset upstreams instead of clear the entire upstreams，I don't know if I've described my problem clearly，

There is another confusion，The deletion granularity of the following two are both individual
```
"balancer:upstreams:" .. id - individual upstreams to be invalidated on individual basis
"balancer:targets:" .. id target for an upstream along with the upstream it belongs to
singletons.cache:invalidate_local("balancer:upstreams:" .. upstream.id)
singletons.cache:invalidate_local("balancer:targets:"   .. upstream.id)
```
Why invalidate the entire upstreams?
`singletons.cache:invalidate_local("balancer:upstreams")`

@bungle @kikito @hbagdi @sabertobihwy

**Comment by liukaiyuliukaiyu** [comment]
I really, really need help, please help answer this question @bungle @kikito @sabertobihwy Thank you very much

**Comment by sabertobihwy** [comment]
> Maybe the version 0.10.3 we're using is older,

This version is too old, so discussing based on this version is meaningless, after all, we won’t do anything on this version.

> rewrite the shared memory, reset upstreams instead of clear the entire upstreams

Reasons for not implementing reset upstreams: 1. The implementation is too complex; 2. There is a risk of memory leaks.

**Reply by liukaiyuliukaiyu** [reply]
Your first point，my implementation logic is as follows
`cache.delete(cache.upstreams_dict_key())	
cache.delete(cache.upstream_key(message_t.entity.id))	
cache.delete(cache.targets_key(message_t.entity.id))
Change to：
cache.add_or_invalidate_individual_upstream_to_upstreams_dict(message_t.type, message_t.entity.name, message_t.entity.id)
cache.delete(cache.upstream_key(message_t.entity.id))
cache.delete(cache.targets_key(message_t.entity.id))

Function add_or_invalidate_individual_upstream_to_upstreams_dict implementation：
local function check_shared_dict_upstreams_updated(events_type, upstream_name, upstream_id)
    local upstreams_dict = _M.sh_get(CACHE_KEYS.UPSTREAMS)
    if upstreams_dict == nil then
        return false
    end
    upstreams_dict = json_decode(upstreams_dict)

    if events_type == "ENTITY_DELETED" then
        if upstreams_dict[upstream_name] == nil then
            return true
        end
    elseif events_type == "ENTITY_CREATED" then
        if upstreams_dict[upstream_name] == upstream_id then
            return true
        end
    end
    return false
end

function _M.add_or_invalidate_individual_upstream_to_upstreams_dict(events_type, upstream_name, upstream_id)
    local upstreams_dict = _M.get(CACHE_KEYS.UPSTREAMS)
    if not upstreams_dict then
        return
    end

    if events_type == "ENTITY_DELETED" then
        for name, id in pairs(upstreams_dict) do
            if name == upstream_name and upstream_id == id then
                upstreams_dict[upstream_name] = nil
                break
            end
        end
    elseif events_type == "ENTITY_CREATED" then
        upstreams_dict[upstream_name] = upstream_id
    end

    DATA:set(CACHE_KEYS.UPSTREAMS, upstreams_dict)

    local lock, err = resty_lock:new("upstreams_cache_locks", {
        exptime = 10,
        timeout = 5
    })
    if not lock then
        log(ngx.ERR, "could not create lock: ", err)
        return
    end
    local elapsed, err = lock:lock(CACHE_KEYS.UPSTREAMS)
    if not elapsed then
        log(ngx.ERR, "failed to acquire cache lock: ", err)
        return
    end

    local res = check_shared_dict_upstreams_updated(events_type, upstream_name, upstream_id)
    if res then
        resty_unlock(lock)
        return
    end
    local _, err = _M.sh_set(CACHE_KEYS.UPSTREAMS, json_encode(upstreams_dict), nil)
    if err then
        log(ngx.ERR, "set shared dict upstreams failed: ", err)
        resty_unlock(lock)
        return
    end
    resty_unlock(lock)
    return
end`
I hope you can criticize and correct when you have time to find the wrong logic
Your second point，Has kong ever encountered memory leak problem? Where can I find this information
Thank you very much for your reply
